t/08_new_api.t

```diff
@@ -0,0 +1,305 @@
+    is $multi->{EMAIL_KEY()},   DEFAULT_EMAIL, 'Default email is returned';
+    is $multi->{REFRESH_KEY()}, DEFAULT_REF,   'Default refresh is returned';
+};
+
+subtest 'Check types' => sub {
+    my $app_config = _new_app_config();
+    is $app_config->get_data_type(NON_EXT_KEY), undef,      'Bad key returns nothing';
+    is $app_config->get_data_type(EMAIL_KEY),   'Str',      'Email type is correct';
+    is $app_config->get_data_type(ADMINS_KEY),  'ArrayRef', 'Admins type is correct';
+    is $app_config->get_data_type(REFRESH_KEY), 'Num',      'Refresh rate type is correct';
+};
+
+subtest 'Check key types' => sub {
+    my $app_config = _new_app_config();
+    is $app_config->get_key_type(NON_EXT_KEY), undef,     'Bad key returns nothing';
+    is $app_config->get_key_type(EMAIL_KEY),   'dynamic', 'Email type is correct';
+    is $app_config->get_key_type(ADMINS_KEY),  'static',  'Admins type is correct';
+    is $app_config->get_key_type(REFRESH_KEY), 'dynamic', 'Refresh rate type is correct';
+};
+
+subtest 'Basic set and get' => sub {
+    my $app_config = _new_app_config();
+
+    ok $app_config->set({EMAIL_KEY() => FIRST_EMAIL}), 'Set 1 value succeeds';
+    is $app_config->get(EMAIL_KEY), FIRST_EMAIL, 'Email is retrieved successfully';
+};
+
+subtest 'Batch set and get' => sub {
+    my $app_config = _new_app_config();
+
+    ok $app_config->set({
+            EMAIL_KEY()   => FIRST_EMAIL,
+            REFRESH_KEY() => REFRESH_SET
+        }
+        ),
+        'Set 2 values succeeds';
+
+    ok my $res = $app_config->get([EMAIL_KEY, REFRESH_KEY]);
+    is $res->{EMAIL_KEY()},   FIRST_EMAIL, 'Email is retrieved successfully';
+    is $res->{REFRESH_KEY()}, REFRESH_SET, 'Refresh is retrieved successfully';
+    exists $res->{'_global_rev'}, 'Batch get returns global revision';
+};
+
+subtest 'Gets/sets of illegal keys' => sub {
+    subtest 'Attempt to set static key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->set({ADMINS_KEY() => ADMINS_SET});
+        }
+        qr/Cannot set with key/;
+    };
+
+    subtest 'Attempt to set non-existent key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->set({NON_EXT_KEY() => ADMINS_SET});
+        }
+        qr/Cannot set with key/;
+    };
+
+    subtest 'Attempt to get non-existent key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->get(NON_EXT_KEY);
+        }
+        qr/Cannot get with key/;
+    };
+};
+
+subtest 'History chronicling' => sub {
+    my $app_config = _new_app_config();
+    my $module     = Test::MockModule->new('Data::Chronicle::Reader');
+
+    subtest 'Add history of values' => sub {
+        # Sleeps ensure the chronicle records them at different times
+        ok $app_config->set({EMAIL_KEY() => FIRST_EMAIL}), 'Set 1st value succeeds';
+        sleep 1;
```

Comment:
`sleep`s in tests always make me a little suspicious. What are these ones for?

Reply:
The resolution of the chronicle timestamp is 1 second, so if we don't wait and call `set` again with the same key then it'll overwrite the history.